readme.md
# makedepend - create dependencies in makefiles


## Introduction

makedepend scans a list of C and C++ source files for `#include` statements,
and then rewrites the makefile to add these header files as "dependencies" of
the target for the C/C++ files. This will cause the Make program to recompile
the C/C++ sources if one of the header files is modified.

Originally, makedepend was distributed as part of the X Window system
(copyright by the Open Group). The version presented here is a derivative of
the original utility. See section [Change List](#Change-List) for details.


## Usage

In a typical case, a makefile contains a pseudo-target to update the dependencies.
The pseudo-target is typically called "depend". Thus, when you `make depend`,
the dependencies for the C/C++ sources get rebuilt.

```make
SRCS   = file1.c file2.c
CFLAGS = -Wall -DDEBUG -I../include

depend :
        makedepend -- $(CFLAGS) $(SRCS)
```

In the above example, makedepend parses the options in the `CFLAGS` macro.
This macro is set after a double hyphen (`--`), so that makedepend will not
warn for any options in `CFLAGS` that it does not recognize. Specifically, in
this example, makedepend handles the `-D` and `-I` options in `CFLAGS` but
silently ignores the `-Wall` option.

By default, makedepend writes its output to a file called `makefile` (on
operating systems with case-sensitive filenames, it tries both `makefile` and
`Makefile`). If it finds it, it appends the generated dependencies to that
makefile. Following the above example, the output of running `make depend` might
become:

```make
SRCS   = file1.c file2.c
CFLAGS = -Wall -DDEBUG -I../include

depend :
        makedepend -- $(CFLAGS) $(SRCS)

# GENERATED DEPENDENCIES. DO NOT DELETE.
file1.o : file1.h file2.h common.h
file2.o : file2.h common.h
```

Running `make depend` again will replace the lines below the comment `# GENERATED DEPENDENCIES...`
with the freshly generated dependency list. This is why it is important to not
remove or change that comment, because if makedepend does not find that line,
it appends its dependency list to the makefile again, rather than replacing the
existing (and possibly outdated) dependency list.

makedepend warns about an include file that it cannot locate, because it will
skip this include file in the dependency list for the target. See also the `-i`
option.

### Keep dependencies separate from build rules

In an environment where you work in a team, and especially when the makefile is
committed to version control, it may be preferable to keep machine-generated
dependencies separate from the general build rules and options. Accordingly,
makedepend is set to write its output to a different file, which is then
included in the makefile.

For example:
```make
SRCS   = file1.c file2.c
CFLAGS = -Wall -DDEBUG -I../include

depend :
        makedepend -fmakefile.dep -- $(CFLAGS) $(SRCS)

-include makefile.dep
```

The `-f` option sets the name of the output file. This output file ("makefile.dep")
is then included in the makefile. The `include` directive (on the last line) is
prefixed by a `-` so that Make won't complain when makefile.dep is initially
missing (this is the GNU Make syntax; other Make's may need a different syntax).

### Automatic updating dependencies

The above examples update the dependencies when making the target `depend`. This
step can also be automated, such that the dependencies are updated when some
targets when the relevant source files change. To this end, the rule for
makedepend must list the source files as the dependencies, and the targets for
the object file must list the generated output file as a dependency.

```make
SRCS   = file1.c file2.c
CFLAGS = -Wall -DDEBUG -I../include

makefile.dep : $(SRCS)
        makedepend -a -f$@ -- $(CFLAGS) $?

$(SRCS:.c=.o) : makefile.dep

-include makefile.dep
```

Note that the shell line for makedepend uses the `$?` macro to list only the
files from `$(SRCS)` on the command line that are newer than the target
`makefile.dep`. This reduces the time needed to update the dependencies, because
only the modified files are scanned.

The other important change on the shell line for makedepend is the `-a` option.
Without this option, the resulting `makefile.dep` would contain only the
dependencies of the files of the most recent run. By default, makedepend removes
all dependency lines from the output file. The `-a` option lets makedepend
remove <em>only</em> the dependency lines of the files on the command line
(which will be regenerated).

The line to add `makefile.dep` as a dependency to all targets that are built
from the list of sources, is GNU Make syntax, using suffix substitution of the
`.c` file extension to `.o`.

### Predefined variables

While makedepend processes the source files, the predefined macro `_makedepend`
is set. Testing for this macro in the source file allows you to to conditionally
exclude a part of the source code from makedepend's processing.

Other predefined variables that makedepend sets are platform-specific. you can
run `makedepend -h -v` to show the list of predefined variables. You can use the
`-U` argument to undefine a predefined variable.

When a source file on the makedepend command line has the extension `.cpp`, `.cxx`,
`.cc` or `.c++`, the variable `__cplusplus` is predefined for that file.

### Locate source files with VPATH

When cross-compiling for a different platform (for example, when developing
embedded software for a microcontroller on a workstation), the system include
paths are likely for the workstation's main compiler setup and may not be
correct for the target compiler. In this case, you can remove the system include
paths with the `-I-` option and set a specific path for "system" includes.

When source files are stored in different directories, with `vpath`'s set to
locate the files, add the source files as dependencies on the `depend` target
line. This way, Make will create the source file list after the `vpath` lookup.

```make
CFLAGS   =-Wall -I./startup/cmsis -std=c1x

CMSIS_O  = bootLPC11Uxx.o systemLPC11Uxx.o
USBHID_O = usbhid.o iap.o tracesupport.o uart.o
COREUSB_O= usb_LPC11Uxx.o usbcore.o usbdescr.o usbuser.o hiduser.o

vpath %.c   ./startup/cmsis     # startup code (CMSIS compatible)
vpath %.c   ./coreusb           # USB support code

.PHONY: depend
depend : $(USBHID_O:.o=.c) $(CMSIS_O:.o=.c) $(COREUSB_O:.o=.c)
        makedepend -fmakefile.dep -b -I- -- $(CFLAGS) $^
```

In the example above, from the top there are the definitions of `CFLAGS` and of
lists of object files for various modules. It is common to list the object files,
because these are the dependencies of the linker command. Following that, are
`vpath` definitions, indicating where Make will search for the C files (this
example assumes GNU Make, the syntax may be different for other Make utilities).

The `depend` target is declared as "phony" for good measure (as it is not a target
that will exist as a file). The dependencies of the `depend` target are the three
lists with the object files, except that the extension `.o` is replaced by `.c`.
This enables you to use the automatic macro `$^` (for "all sources") on the
`makedepend` command line. When Make builds the file list for `$^`, it looks
through the `vpath` directories for the files, and it will add the directories
to the files where applicable. For example, if `bootLPC11Uxx.c` is found in
`./startup/cmsis/`, the full path `./startup/cmsis/bootLPC11Uxx.c` is included
in `$^`.


## Options
<dl>
<dt> <code>-D</code>name=def </dt>
<dt> <code>-D</code>name </dt>
<dd>
  Defines a symbol for the makedepend preprocessor (which works like the
  C/C++ preprocessor). When no value is explicitly defined, the symbol is
  defined as "1".
</dd>

<dt> <code>-I</code>path </dt>
<dd>
  Adds a path to its list of directories that makedepend searches for
  files when it encounters a <code>#include</code> directive.
  <br><br>
  By default, makedepend appends the standard include directories at the end
  of the directory list. In Linux, Unix and OS/2, makedepend evaluates the
  <code>C_INCLUDE_PATH</code> environment variable for the standard includes; in
  Microsoft Windows, makedepend evaluates the <code>INCLUDE</code> environment
  variable. When the option <code>-I-</code> is set, makedepend does <em>not</em>
  append the standard include directories (and thus prevents makedepend from
  searching the standard include directories).
</dd>

<dt> <code>-U</code>name </dt>
<dd>
  Undefines a symbol. This is mostly useful to remove a predefined symbol.
</dd>

<dt> <code>-a</code> </dt>
<dd>
  Accumulates the dependencies in the output file instead of removing the
  dependencies for files that are <em>not</em> listed on the command line of
  makedepend. With this option, you can call makedepend multiple times with
  different filename lists, and obtaining the accumulated dependencies of all
  those calls.
  <br><br>
  Note that dependencies for files that are listed on the command line are
  replaced in the output file. The <code>-a</code> option is not a simple
  "append".
</dd>

<dt> <code>-b</code> </dt>
<dd>
  No backups. By default, makedepend copies the makefile to one with a .bak
  extension before modifying it. When this option is set, makedepend deletes
  the backup file.
</dd>

<dt> <code>-c</code> </dt>
<dd>
  Includes the C/C++ source file in the list of dependencies. By default,
  makedepend only lists all files <em>included</em> by the source file on the
  dependency line.
</dd>

<dt> <code>-e</code> </dt>
<dd>
  Excludes <em>system includes</em> from the list of dependencies. System
  includes are files that are between angle brackets (after the #include).
  These are less relevant to list as dependecies; the system include files
  can be considered stable.
</dd>

<dt> <code>-f</code>makefile </dt>
<dd>
  Sets the output filename, instead of the default name "makefile". When set
  to "-" (i.e. option <code>-f-</code>), the output is sent to standard output
  (the console).
</dd>

<dt> <code>-h</code> </dt>
<dd>
  Help. Shows brief usage information for makedepend. When verbose output is
  enabled (see <code>-v</code>), the list of predefined variables is listed too.
</dd>

<dt> <code>-i</code> </dt>
<dd>
  Ignore include files that cannot be located. That is, do <b>not</b> warn about any
  include files that cannot be found. Note that include files that are not
  found, are <i>never</i> added as a dependency to the target. This option silences
  the warning about missing files.
</dd>

<dt> <code>-m</code> </dt>
<dd>
  Enables warnings for multiple inclusion (of the same file). This option is
  provided to aid in debugging problems related to multiple inclusion.
</dd>

<dt> <code>-o</code>suffix </dt>
<dd>
  Object file suffix. The default suffix is ".o", as is common for Unix-like
  operating systems and the GNU GCC compiler suite. Om Microsoft Windows, the
  suffix ".obj" is more common, which can be set with the option <code>-o.obj</code>.
</dd>

<dt> <code>-p</code>prefix </dt>
<dd>
  Sets the text that is prepended to the name of each object file. This is
  usually used to designate a different directory for the object file.
  <br><br>
  If the prefix pattern starts with a minus ("-"), that prefix is <em>removed</em>
  from the object filename (if the object filename starts with that prefix).
  <br><br>
  The default prefix string is the empty string.
</dd>

<dt> <code>-s</code>string </dt>
<dd>
  Sets the delimiter string (or "separator") below which makedepend writes the
  generated depedencies. The default string is
  "#&nbsp;GENERATED&nbsp;DEPENDENCIES.&nbsp;DO&nbsp;NOT&nbsp;DELETE.".
</dd>

<dt> <code>-v</code> </dt>
<dd>
  Verbose operation: this option causes makedepend to show more notices and
  warnings during processing the files. It will also emit the list of files
  included by each input file.
</dd>

<dt> <code>-w</code>width </dt>
<dd>
  Sets the maximum line width for the lines written to the output file. The
  default value is 78 characters.
</dd>

<dt> <code>-include</code> file </dt>
<dd>
  Processes the file includes it before processing each regular input file.
  This has the same affect as if the specified file were listed in an
  <code>#include</code> directive at the very to of each regular input file.
  <br><br>
  The <code>-include</code> option may be present multiple times on the command
  line.
</dd>

<dt> <code>--</code> options <code>--</code> </dt>
<dd>
  Following a double hyphen (<code>--</code>), only the <code>-D</code>, <code>-I</code> and <code>-U</code> options are
  handled, and any others are silently ignored (source filenames are still
  handled). The intended purpose of this syntax is that the same <code>CFLAGS</code>
  macro that is passed to the C/C++ compiler can also be passed to makedepend.
  <br><br>
  A second double hyphen ends this special processing; it is needed if options
  that are specific to makedepend follow the definitions in <code>CFLAGS</code> (or
  another makefile macro).
</dd>
</dl>


## License

Copyright (c) 1993, 1994, 1998 The Open Group

Permission to use, copy, modify, distribute, and sell this software and its
documentation for any purpose is hereby granted without fee, provided that the
above copyright notice appear in all copies and that both that copyright notice
and this permission notice appear in supporting documentation.

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
THE OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of The Open Group shall not be
used in advertising or otherwise to promote the sale, use or other dealings
in this Software without prior written authorization from The Open Group.


## Change List

This version is derived from a port of makedepend by [Derell Licht](https://github.com/DerellLicht/makedepend) (circa 2005)
of the original source code to a contemporary version of GNU GCC. It has since
been modified to bring new features and improvements:

* Some minor adjustments for better compatibility with Microsoft Windows, such
  as handling the backslash as an equivalent as the forward slash in paths.
* Filenames with space characters are written with the space characters
  "escaped" with a backslash (in a way compatible with GNU Make).
* The `-a` command line option now does an intelligent "append", where
  dependencies for new files are added, but if a file that makedepend scans
  already appears in the section below the delimiter, the dependencies for that
  file are replaced.
* Added the `-b` command line option, to skip making backups of the input
  makefiles.
* Added the `-c` command line option, to add the source file as a dependency
  for each target (in addition to the files included by that source file).
* Added the `-e` command line option, to list only project-specific include
  files as dependencies (and skip any system includes). The rationale is that
  this option makes a Makefile more portable, because all listed dependencies
  are part of the project (all system-specific dependencies are excluded).
* Added the `-h` command line option, to show program usage. The `-?` option is
  equivalent to `-h`.
* Let makedepend display the (platform-specific) list of predefined variables
  with the `-h -v` arguments (verbose help).
* Replace the option `-Y` by `-I-`.
* Let makedepend create the "makefile" with dependencies if it does not yet
  exist.
* Read the `INCLUDE` environment variable on Microsoft Windows (instead of the
  `C_INCLUDE_PATH` variable).
* Added the prededined `_makedepend` and `__cplusplus` variables.
* Add a warning for the case that an include file cannot be located (because
  that file will then not be added as a dependency). The `-i` option silences
  the warning.


ChangeLog
github commit 080b4e4533c60b0ce2809242afa6bb1a324be0b1
Author: Thiadmer Riemersma
Date:   Wednesday 9 February 2022

    makedepend 1.0.8

    Added a warning for include files that cannot be located (as these are
    skipped in the dependency list). Added the -i option to silence the warning.


github commit 2852bd0500f1eec159ed00a69ef3df803bf087c4
Author: Thiadmer Riemersma
Date:   Sunday 11 October 2020

    makedepend 1.0.7

    Added the -e command line option, to exclude "system" include files on
    the dependency lists.

    Handle each source file only once, even if it appears multiple times on
    the command line.


github commit 2597250be5a067f5a8caa67a63e4e8aba19464cc
Author: Thiadmer Riemersma
Date:   Sunday 25 October 2019

    makedepend 1.0.6

    Fix for a memory allocation bug.

    Fixes for portability to Linux.


github commit a4a8449910e18b014550003ebc7668abd8a45051
Author: Thiadmer Riemersma
Date:   Monday 13 May 2019

    Some minor adjustments for better compatibility with Microsoft Windows,
    such as handling the backslash as an equivalent as the forward slash in
    paths.

    Filenames with space characters are written with the space characters
    "escaped" with a backslash (in a way compatible with GNU Make).

    The -a command line option now does an intelligent "append", where
    dependencies for new files are added, but dependencies for existing files
    are replaced.

    Added the -b command line option, to skip making backups of the input
    makefiles.

    Added the -c command line option, to add the source file as a dependency
    for each target (in addition to the files included by that source file).

    Added the -h command line option, to show program usage.

    Let makedepend display the (platform-specific) list of predefined variables
    with the -h -v arguments (verbose help).

    Replace the option -Y by -I-.

    Let makedepend create the "makefile" with dependencies if it does not yet
    exist.

    Read the INCLUDE environment variable on Microsoft Windows (instead of
    the C_INCLUDE_PATH variable).

    Added the prededined _makedepend and __cplusplus variables.


commit 30859adfd6a6523a92a637ca26bd61ce59680a1b
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Mon Jul 22 22:14:36 2013 -0700

    makedepend 1.0.5

    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>

commit 5b6d3797eae3bac824cf72c0c8f64f51e3d4eaef
Author: Julien Cristau <jcristau@debian.org>
Date:   Mon Jul 1 18:57:12 2013 +0200

    Don't crash if an option's argument is missing

    Avoid dereferencing argv[argc].

    See http://www.forallsecure.com/bug-reports/011f1a55f79a5501b36008d6ee0d40e8b6644569/

    Reported-by: Alexandre Rebert <alexandre@cmu.edu>
    Signed-off-by: Julien Cristau <jcristau@debian.org>
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>

commit 107c2fcf5c7576ee66ee23926f0431b6db27a07a
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Sat Feb 23 09:57:51 2013 -0800

    Assume signal handlers return void, as C89 requires

    Drops use of autoconf's obsolete AC_TYPE_SIGNAL

    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>

commit b2c6d0d0f64d419590c7cd77aae4890866665cba
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Tue Feb 12 18:36:30 2013 -0800

    Bug 56091 - Unsafe use of strcpy() in makedepend

    Use memmove for potentially overlapping copies.

    Reported-by: Laurence Jupp <laurence@narya.org>
    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Reviewed-by: Matthieu Herrb <matthieu.herrb@laas.fr>

commit 897e25de0cf27bd345130912a6da34b0c802a0ca
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Tue Feb 12 18:33:44 2013 -0800

    Store len in a size_t instead of an int in redirect()

    Fixes clang warnings about converting back & forth:

    main.c:761:8: warning: implicit conversion loses integer precision: 'size_t'
          (aka 'unsigned long') to 'int' [-Wshorten-64-to-32]
            len = strlen(line);
                ~ ^~~~~~~~~~~~
    main.c:763:41: warning: implicit conversion changes signedness: 'int' to
          'size_t' (aka 'unsigned long') [-Wsign-conversion]
                    if (*buf == '#' && strncmp(line, buf, len) == 0)
                                       ~~~~~~~            ^~~

    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>

commit 87fd28cd5a5acfe6e91c97c4af0a469e1e86cbd5
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Sun Jan 20 22:45:53 2013 -0800

    Mark fatalerr & catch as _X_NORETURN

    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>

commit 637c59e78f699a824e78316cb8ca3113bb5c738b
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Sat Jan 12 14:40:40 2013 -0800

    Replace deprecated Automake INCLUDES variable with AM_CPPFLAGS

    Excerpt https://lists.gnu.org/archive/html/automake/2012-12/msg00038.html

      - Support for the long-deprecated INCLUDES variable will be removed
        altogether in Automake 1.14.  The AM_CPPFLAGS variable should be
        used instead.

    This variable was deprecated in Automake releases prior to 1.10, which is
    the current minimum level required to build X.

    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Reviewed-by: Julien Cristau <jcristau@debian.org>

commit 059e9be5a7321974b190d786d426b2ae27be8c56
Author: Glenn Burkhardt <gbburkhardt@gmail.com>
Date:   Fri Jan 11 22:09:54 2013 -0800

    Bug 52099 - makedepend rename fails on CIFS mounted shares

    Patch v1: https://bugs.freedesktop.org/attachment.cgi?id=64218

    v2: modified by Alan Coopersmith to restore Makefile to original name
    if contents cannot be read, instead of leaving it moved to .bak name.

    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>

commit bb29c304c09aba567af0f0d77a63f0202c3836c6
Author: Colin Walters <walters@verbum.org>

def.h
/* $Xorg: def.h,v 1.4 2001/02/09 02:03:16 xorgcvs Exp $ */
/*

Copyright (c) 1993, 1994, 1998 The Open Group.

Permission to use, copy, modify, distribute, and sell this software and its
documentation for any purpose is hereby granted without fee, provided that
the above copyright notice appear in all copies and that both that
copyright notice and this permission notice appear in supporting
documentation.

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of The Open Group shall not be
used in advertising or otherwise to promote the sale, use or other dealings
in this Software without prior written authorization from The Open Group.

*/
/* $XFree86: xc/config/makedepend/def.h,v 3.15tsi Exp $ */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <assert.h>
#if !defined _MSC_VER
  #include <unistd.h>
#endif
#if 0
  #ifndef X_NOT_POSIX
    #ifndef _POSIX_SOURCE
    #define _POSIX_SOURCE
    #endif
  #endif
#endif
#include <sys/types.h>
#include <fcntl.h>
#include <sys/stat.h>

#if defined _MSC_VER
  #define snprintf    _snprintf
#endif

#define MAXDEFINES  512
#define MAXFILES    2048
#define MAXINCFILES 128 /* "-include" files */
#define MAXDIRS     64
#define SYMTABINC   10  /* must be > 1 for define() to work right */
#define TRUE        1
#define FALSE       0

/* the following must match the directives table in main.c */
#define IF              0
#define IFDEF           1
#define IFNDEF          2
#define ELSE            3
#define ENDIF           4
#define DEFINE          5
#define UNDEF           6
#define INCLUDE         7   /* system include, using <...> */
#define LINE            8
#define PRAGMA          9
#define ERROR           10
#define IDENT           11
#define SCCS            12
#define ELIF            13
#define EJECT           14
#define WARNING         15
#define INCLUDENEXT     16  /* system include_next, using <...> */
#define IFFALSE         17  /* pseudo value --- never matched */
#define ELIFFALSE       18  /* pseudo value --- never matched */
#define INCLUDEUSR      19  /* pseudo value --- never matched (non-system #include, using "...") */
#define IFGUESSFALSE    20  /* pseudo value --- never matched */
#define ELIFGUESSFALSE  21  /* pseudo value --- never matched */
#define INCLUDENEXTUSR  22  /* pseudo value --- never matched (non-system #include_next, using "..." */

#ifdef DEBUG
  extern int  _debugmask;
  /*
   * debug levels are:
   *
   *     0    show ifn*(def)*,endif
   *     1    trace defined/!defined
   *     2    show #include
   *     3    show #include SYMBOL
   *     4-6  unused
   */
  #define debug(level,arg) { if (_debugmask & (1 << level)) warning arg; }
#else
  #define debug(level,arg) /**/
#endif /* DEBUG */

typedef unsigned char boolean;

struct symtab {
  const char *s_name;
  const char *s_value;
  const char *s_args;
};

/* possible i_flag */
#define DEFCHECKED  (1<<0)  /* whether defines have been checked */
#define NOTIFIED    (1<<1)  /* whether we have revealed includes */
#define MARKED      (1<<2)  /* whether it's in the makefile */
#define SEARCHED    (1<<3)  /* whether we have read this */
#define FINISHED    (1<<4)  /* whether we are done reading this */
#define INCLUDED_SYM    (1<<5)  /* whether #include SYMBOL was found
                                 * Can't use i_list if TRUE */
#define FORCED_DEP  (1<<6)  /* file is forced as a dependency */

struct  inclist {
  char   *i_incstring;      /* string from #include line */
  char   *i_file;           /* path name of the include file */
  struct inclist **i_list;  /* list of files it itself includes */
  int    i_listlen;         /* length of i_list */
  struct symtab **i_defs;   /* symbol table for this file and its children
                             * when merged */
  int     i_ndefs;          /* current # defines */
  boolean *i_merged;        /* whether we have merged child defines */
  unsigned char i_flags;
};

struct filepointer {
  const char *f_name;
  char    *f_p;
  char    *f_base;
  char    *f_end;
  long    f_len;
  long    f_line;
  long    cmdinc_count;
  char    **cmdinc_list;
  long    cmdinc_line;
};

#include <stdlib.h>
#if defined(macII) && !defined(__STDC__)  /* stdlib.h fails to define these */
  char *malloc(), *realloc();
#endif /* macII */

char*  copy(const char *str);
int    match(const char *str, const char **list);
char*  getnextline(struct filepointer *fp);
char*  base_name (const char *file);
struct symtab** slookup(const char *symbol, struct inclist *file);
struct symtab** isdefined(const char *symbol, struct inclist *file,
                          struct inclist **srcfile);
struct symtab** fdefined(const char *symbol, struct inclist *file,
                         struct inclist **srcfile);
struct filepointer* getfile(const char *file);
void   included_by(struct inclist *ip, struct inclist *newfile);
struct inclist* newinclude(char *newfile, const char *incstring);
void   inc_clean(void);
struct inclist* inc_path(const char *file, char *include, int type);

void   freefile(struct filepointer *fp);

void   define2(const char *name, const char *args, const char *val,
               struct inclist *file);
void   define(char *def, struct inclist *file);
void   undefine(const char *symbol, struct inclist *file);
int    find_includes(struct filepointer *filep,
                     struct inclist *file,
                     struct inclist *file_red,
                     int recursion, boolean failOK);

char*  targetname(const char *base);
void   recursive_pr_include(struct inclist *head, const char *file);

int    cppsetup(const char *filename,
                const char *line,
                struct filepointer *filep,
                struct inclist *inc);

void   memoryerr(void);
void   fatalerr(const char*, ...);
void   warning(const char*, ...);
void   warning1(const char*, ...);
void   showusage(void);

/* defined in main.c */
extern const char * const directives[];
